### ISSUE TITLE
Transitive dependency linking back to workspace shows in `pnpm why`, but is not installed nor included in `pnpm ls`

### ISSUE THREAD

**Issue by jakebailey** [pnpm/pnpm#7283]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

8.10.2

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/51f33ee5c67eadf5392a64d5b196b4c722fc3c64

### Reproduction steps

```console
$ pnpm why --filter ./types/express-socket.io-session @types/engine.io        
Legend: production dependency, optional only, dev only

@types/express-socket.io-session@1.3.9999 /home/jabaile/work/DefinitelyTyped/types/express-socket.io-session

dependencies:
@types/socket.io 2.1.13
└── @types/engine.io link:../engine.io

devDependencies:
@types/express-socket.io-session link:
└─┬ @types/socket.io 2.1.13
  └── @types/engine.io link:../engine.io

$ pnpm install --filter . --filter '...{./types/express-socket.io-session}...'
```

```console
$ ls types/engine.io 
engine.io-tests.ts  index.d.ts  package.json  tsconfig.json  tslint.json
$ pnpm test express-socket.io-session

> definitely-typed@0.0.3 test /home/jabaile/work/DefinitelyTyped
> node --enable-source-maps node_modules/@definitelytyped/dtslint/ types "express-socket.io-session"

dtslint@0.0.190
Error: Errors in typescript@5.4 for external dependencies:
../engine.io/index.d.ts(1,23): error TS2688: Cannot find type definition file for 'node'.

    at testTypesVersion (/home/jabaile/work/DefinitelyTyped/node_modules/.pnpm/@definitelytyped+dtslint@0.0.190_typescript@5.4.0-dev.20231106/node_modules/@definitelytyped/dtslint/src/index.ts:202:11)
    at runTests (/home/jabaile/work/DefinitelyTyped/node_modules/.pnpm/@definitelytyped+dtslint@0.0.190_typescript@5.4.0-dev.20231106/node_modules/@definitelytyped/dtslint/src/index.ts:170:7)
    at main (/home/jabaile/work/DefinitelyTyped/node_modules/.pnpm/@definitelytyped+dtslint@0.0.190_typescript@5.4.0-dev.20231106/node_modules/@definitelytyped/dtslint/src/index.ts:83:5)
 ELIFECYCLE  Test failed. See above for more details.
```

```console
$ pnpm ls --depth -1 --filter '...{./types/express-socket.io-session}...' --parseable | sort
/home/jabaile/work/DefinitelyTyped/types/body-parser
/home/jabaile/work/DefinitelyTyped/types/connect
/home/jabaile/work/DefinitelyTyped/types/cookie-parser
/home/jabaile/work/DefinitelyTyped/types/empower
/home/jabaile/work/DefinitelyTyped/types/empower-core
/home/jabaile/work/DefinitelyTyped/types/express
/home/jabaile/work/DefinitelyTyped/types/express-serve-static-core
/home/jabaile/work/DefinitelyTyped/types/express-session
/home/jabaile/work/DefinitelyTyped/types/express-socket.io-session
/home/jabaile/work/DefinitelyTyped/types/http-errors
/home/jabaile/work/DefinitelyTyped/types/mime
/home/jabaile/work/DefinitelyTyped/types/mime/v1
/home/jabaile/work/DefinitelyTyped/types/node
/home/jabaile/work/DefinitelyTyped/types/power-assert
/home/jabaile/work/DefinitelyTyped/types/power-assert-formatter
/home/jabaile/work/DefinitelyTyped/types/qs
/home/jabaile/work/DefinitelyTyped/types/range-parser
/home/jabaile/work/DefinitelyTyped/types/send
/home/jabaile/work/DefinitelyTyped/types/serve-static
```

### Describe the Bug

`@types/engine.io` correctly links to the workspace when shown in `pnpm why`, but `pnpm install` doesn't install it. It's not included in the list of packages when doing `pnpm ls --depth -1` either.

### Expected Behavior

`./types/engine.io` is installed when the filter is `'...{./types/express-socket.io-session}...'` and it appears in `ls`, as it does with `pnpm why`.

See also https://github.com/DefinitelyTyped/DefinitelyTyped/actions/runs/6774736687/job/18412478604?pr=67331 and https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67331. DT is currently failing in CI when we use partial mode for some packages.

### Which Node.js version are you using?

20.9.0

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [X] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

Linux 5.15.90.1-microsoft-standard-WSL2 x86_64 x86_64

**Comment by KSXGitHub** [comment]
> but `pnpm install` doesn't install it.

But `pnpm install` did install it?

```text
$ stat types/express-socket.io-session/node_modules/@types/socket.io          
  File: types/express-socket.io-session/node_modules/@types/socket.io -> ../.pnpm/@types+socket.io@2.1.13/node_modules/@types/socket.io
  Size: 62        	Blocks: 8          IO Block: 4096   symbolic link
Device: 0,44	Inode: 90444773    Links: 1
```

**Comment by jakebailey** [comment]
See my `ls` above; it did not get a node_modules dir so its deps were not installed.

**Comment by jakebailey** [comment]
To be clear I mean "installed" in the sense of "what you get if you were to cd to that dir and `pnpm install`.

**Comment by KSXGitHub** [comment]
This is the content of `engine.io/package.json`:

```json
{
    "private": true,
    "name": "@types/engine.io",
    "version": "3.1.9999",
    "projects": [
        "https://github.com/socketio/engine.io"
    ],
    "dependencies": {
        "@types/node": "*"
    },
    "devDependencies": {
        "@types/engine.io": "workspace:."
    },
    "owners": [
        {
            "name": "KentarouTakeda",
            "githubUsername": "KentarouTakeda"
        }
    ]
}
```

There are no other dependencies than `@types/node`. Though, I'm not sure how `pnpm` should handle `@types/node`.

**Comment by jakebailey** [comment]
It should be linking that to the package in the repo, which it does normally, just not in this package when filtered in this specific way.

**Comment by jakebailey** [comment]
Curiously, once things are installed, you can use an infinite depth instead and find it:

```console
$ pnpm ls --depth Infinity --filter '...{./types/express-socket.io-session}...' --parseable | grep -v node_modules | sort -u
/home/jabaile/work/DefinitelyTyped/types/cookie-parser
/home/jabaile/work/DefinitelyTyped/types/engine.io
/home/jabaile/work/DefinitelyTyped/types/express
/home/jabaile/work/DefinitelyTyped/types/express-session
/home/jabaile/work/DefinitelyTyped/types/express-socket.io-session
/home/jabaile/work/DefinitelyTyped/types/node
```

I think this means I can do a temporary workaround whereby I do the filtered install, then run the above, and if any of those aren't installed, install them too.

**Comment by gluxon** [comment]
This one might be related https://github.com/pnpm/pnpm/issues/7378?

**Comment by zkochan** [comment]
Currently, when calculating relationships between workspace projects (dependents/dependencies), we don't take into account the scenario in which a workspace project is a dependency of a package from the registry (I think it is done by the `@pnpm/workspace.pkgs-graph` package).

The easiest way to fix this would probably be to:

1. if install runs with `--filter` and the lockfile is not up-to-date, make a full `lockfile-only` install first with all projects included.
2. with an up-to-date fully resolved lockfile we can properly detect all relations between workspace projects. Workspace filtering will have to be rewritten because right now we only use information from the `package.json` files to detect relations between projects. We'll have to use information from the lockfile instead to detect the more complex deep relations.
3. after this we can proceed to the regular install command (or another command).

So, in a nutshell, filtering will only work with a fully resolved dependency graph.

**Comment by jakebailey** [comment]
The tricky bit is that DT purposefully doesn't use a lockfile.

(And technically, there are cases where not having a lockfile is not enough to get pnpm to pull from npm when we want, i.e. a reinstall doesn't check the registry for new versions so we have to clean and reinstall when dtslint updates)

**Comment by zkochan** [comment]
So add the lockfile to `.gitignore`, I guess.

**Comment by jakebailey** [comment]
We also use `shared-workspace-lockfile=false`, so that may throw a thorn in this too.

**Comment by zkochan** [comment]
Closing for lack of a reproduction on a supported version.

The repository state this came from is long gone. Please open a new issue with a reproduction if `pnpm why` still reports a dependency that is not installed.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
